c/src/space.rs

```diff
@@ -1,12 +1,376 @@
+/// 
+/// subst
+///   \arg payload \c is the pointer to the space's payload
+///   \arg pattern \c is the pattern atom to match.  This function should NOT take ownership of the pattern atom.
+///   \arg tmpl \c is the template atom.  This function should NOT take ownership of the template atom.
+///   NOTE: If a subst function is provided, it will be called.  If NULL is provided, the default
+///     implementation will be called.
+/// 
+/// add
+///   \arg payload \c is the pointer to the space's payload
+///   \arg atom \c is the atom to add to the space.  This function SHOULD take ownership of the atom.
+/// 
+/// remove
+///   \arg payload \c is the pointer to the space's payload
+///   \arg atom \c is the atom to remove from the space.  This function should NOT take ownership of the atom.
+///   Returns `true` if the atom was removed, otherwise returns `false`
+/// 
+/// replace
+///   \arg payload \c is the pointer to the space's payload
+///   \arg from \c is the atom to replace in the space.  This function should NOT take ownership of the `from` atom.
+///   \arg to \c is the atom to replace it with.  This function SHOULD take ownership of the `to` atom.
+///   Returns `true` if the atom was replaced, otherwise returns `false`
+/// 
+/// free_payload
+///   \arg payload \c is the pointer to the space's payload
+///   NOTE: This function is responsible for freeing the payload buffer, as well as any other objects
+///   and resources stored by the space.  This includes `atom_t` objects, `space_observer_t` objects,
+///   as well as any other buffers stored in the space
+#[repr(C)]
+pub struct space_api_t {
+    register_observer: extern "C" fn(payload: *mut c_void, observer: *mut space_observer_t),
+
+    query: extern "C" fn(payload: *mut c_void, atom: *const atom_t) -> *mut bindings_set_t,
+
+    subst: Option<extern "C" fn(payload: *mut c_void, pattern: *const atom_t, tmpl: *const atom_t) -> *mut vec_atom_t>,
+
+    add: extern "C" fn(payload: *mut c_void, atom: *mut atom_t),
+
+    remove: extern "C" fn(payload: *mut c_void, atom: *const atom_t) -> bool,
+
+    replace: extern "C" fn(payload: *mut c_void, from: *const atom_t, to: *mut atom_t) -> bool,
+
+    free_payload: extern "C" fn(payload: *mut c_void),
+}
+
+//QUESTION FOR VITALY: Looking to the CGrounded type for guidance about your preferred
+// patterns, I see that the type CGrounded wraps an AtomicPtr<>, but I don't understand
+// the situation where multiple threads would have shared access to this pointer and need
+// to atomically update it
+struct CSpace {
+    api: *const space_api_t,
+    payload: *mut c_void,
+}
+
+impl Space for CSpace {
+    fn register_observer(&self, observer: Rc<RefCell<dyn SpaceObserver>>) {
+        let api = unsafe{ &*self.api };
+        let observer_ptr = Box::into_raw(Box::new(space_observer_t{observer}));
+        (api.register_observer)(self.payload, observer_ptr);
+    }
```

Comment:
I believe there are three separate forks to this discussion:
1.) I don't understand the problem you are trying to solve with the suggestion above.

I believe the current implementation works.  It doesn't leak resources, as long as all space implementations follow the contract of freeing all registered observers in their own `free_payload()` function implementations.

This is because `Rc` will decrement its reference counter when it is dropped.  The Rc that's passed into `Space::register_observer` gets boxed and converted into a raw ptr which is given to the C `register_observer` callback.  Then eventually the C side will call `space_observer_free`, causing the box to be reconstructed and dropped, leading to the dropping of the Rc, and hence the ref-counter decrement.

I don't think it's necessary or desirable in this case to downgrade the Rc to a Weak.  In my experience, Weak is most useful in situations where circular references are necessary.  Perhaps I am missing something.

2.) What is the best API in C?

I wanted to expose the C `register_observer` callback function to take a simple function ptr (lambda), and not include the space_observer_t object at all, or perhaps, as you say, have a simple (non-opaque) struct that bundles together a function-ptr and a `void* context`.  My reason was that it would be a little bit simpler for the C programmer.

Because of the fact that it's a trait object with an Rc, we have to keep this type opaque.  If we want to get rid of the Rc from this interface, we can simplify it on the C side.  However, that will require some refactoring on the Rust side.  

Ultimately we could go to a simple borrow on the Rust side (instead of an Rc), but that would force the observers to live as long as the space, which will be a pain.  

My favorite alternative is to give the observer ownership to the space, and have the ability to borrow an observer back when you need to inspect it.  For example:
```
pub struct ObserverId(usize);

impl Space for MySpace {
    fn register_observer(&self, observer: Box<dyn SpaceObserver>) -> ObserverId {
        ...
    }
    fn get_observer(&self, obs_id: ObserverId) -> &dyn SpaceObserver {
        ...
    }
}

2.) Do we want a `unregister_observer` function?

Since I didn't see anything like that in Rust, my assumption was that observers just get unregistered (freed) when the space itself is freed.  However, if a space is a long-lived object, and observers will come and go, them perhaps we need an API to unregister them.

Reply:
Ah!  I realized that that the reason behind using an Rc<> for the argument in Space::register_observer() was that you want the observer to automatically unregister from the space when it is dropped by the caller that added it.

So I infer that an observer is potentially a short-lived object, but a space is a longer-lived object.  I hadn't appreciated from the earlier comments that automatic unregistration prior to the space being freed was important.

Just as C lacks automatic freeing for memory, we will need an explicit unregister mechanism here.  Perhaps I can make space_observer_free() trigger unregistration to remain orthogonal with the Rust API.

Reply:
Luke, you last comment is absolutely correct: I have introduced `Weak`s and `Rc`s to implement an automatic deregisteration of  observers when they are dropped by the owning side. I believe this answers question in item (1).

But automatic deregistration is not a "must have" requirement. Here I also experimented with Rust and tried to figure out which interface will be natural. I should say I don't like the complexities the current implementation impose on C API and I would vote to change it into what you have suggested in item (2): `register_observer(&self, observer: Box<dyn SpaceObserver>) -> ObserverId`.

This will naturally lead us to having the `unregister_observer` function in Rust (this anwers the question in the item (3)).

Reply:
I agree moving a collections of  observers into a Rust proxy simplifies the implementations of the C space. I think it is a good decision. I also thought about something like this for the Rust implementation, but didn't do it. If you want to do it as well please raise a separate PR - this one is big enough already ;-).

I am still not sure if usage of `Weak` pointers makes sense. Maybe passing the ownership to the space would be much simpler. I would experiment with it but it should be separate PR for sure.

Reply:
It's a question of whether the burden of manual unregistration is a big deal or not.  And the answer to that question depends on how far apart (in time and in code) the places are where the observer is registered and the place where it's known that it's no longer needed.

I admit that I didn't understand the purpose behind your design at first - passing the Rc<> - but now that I understand how observers are used and their life-cycle with respect to spaces, I think your design that supports automatic unregistration appears to be a very good solution.

Reply:
Added this thread as a comment to https://github.com/trueagi-io/hyperon-experimental/issues/287#issuecomment-1535686844